[export] Use bytecode to flatten graph inputs.

zhxchen17 opened the pull request:
Summary:
Previously for dynamo_graph_capture_for_export, we use make_fx() to trace through the dynamo bytecode so that the GraphModule looks like the following:
```
tree_leaves = pytree_flatten(*args)
graph_inputs = self.in_shuffle_graph(*tree_leaves)
graph_outputs = graph(*graph_inputs)
tree_outs = outputs = self.out_shuffle_graph(*graph_outputs)
return pytree.tree_unflatten(tree_outs)
```

After a bunch of experiments, we think the make_fx step is not necessaery, and to further close the gap between strict export and dynamo, we propose the new impl to do the following using torch.fx codegen:

```
graph_inputs = self._dynamo_bytecode_flatten(*args)
graph_outs = graph(*graph_inputs)
return self._dynamo_bytecode_unflatten(*graph_outs)
```
The `_dynamo_bytecode_flatten` and `_dynamo_bytecode_unflatten` wrappers are constructed in the way that the behavior will be 100% identical to directly running bytecode.

This is important for also making precompile reusing the same piece of logic that torch.export uses.

Test Plan: test:test_export -- -r _strict_export_v2

Differential Revision: D89673950




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan commented on torch/_dynamo/functional_export.py:
```diff
@@ -618,6 +619,235 @@ def normalize_graph_module(gm):
+        return args, kwargs
+
+
+def make_fx_graph_from_captured_output(
```
We are not make_fx-ing anymore right?

zhxchen17 replied:
no in the new impl there's no make_fx()

tugsbayasgalan commented on torch/_dynamo/functional_export.py:
```diff
@@ -618,6 +619,235 @@ def normalize_graph_module(gm):
+    class Yield(Exception):
+        pass
+
+    class DynamoBytecodeFlatten:
```
Should we define them outside of the function definition?

tugsbayasgalan commented on torch/_dynamo/functional_export.py:
```diff
@@ -618,6 +619,235 @@ def normalize_graph_module(gm):
+            assert graph_module._wrapped_call.cls_call is None
+
+            def dynamo_wrapped_call(self, *args, **kwargs):
+                assert "forward" not in self.__dict__
```
Does this mean graph_module will always have class forward implementaiton, not the instance specific one?

zhxchen17 replied:
Yeah, in the current torch.fx impl, we will always use class's forward. I think in real world it'd be weird to patch a generated GraphModule's forward in any sense... but we can add support for that if it's needed in the future.

tugsbayasgalan commented on torch/_dynamo/functional_export.py:
```diff
@@ -618,6 +619,235 @@ def normalize_graph_module(gm):
+                    self._forward_pre_hooks = fwd_pre_hooks
+                    return original_forward(self, *args, **kwargs)
+
+                try:
```
Can you pls add inline comments what is happening here?

tugsbayasgalan replied:
Are you just trying to preserve the forward hooks because for some reason dynamo doesn't trace them?

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -277,24 +278,32 @@ def _extract_fake_inputs(gm, args, kwargs):
+
+        def mark_inputs(x):
+            nonlocal idx
+            fake_inps.append(x)
```
Should we check if x is fake tensor beforehand?

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -277,24 +278,32 @@ def _extract_fake_inputs(gm, args, kwargs):
-        ):
-            if node in node_map:
-                new_fake_inps.append(fake_inps[node_map[node]])
+    if dynamo_bytecode_flatten := getattr(gm, "_dynamo_bytecode_flatten", None):
```
Wait why do we even need this? Can't we just use fake mode from dynamo directly refakify the inputs? So that we can avoid this whole block?

zhxchen17 replied:
> Can't we just use fake mode from dynamo directly refakify the inputs? So that we can avoid this whole block?

Again, fake tensor can be handled this way, but this approach doesn't scale to symint inputs.

tugsbayasgalan reviewed (commented):
Mostly looks good! But would still like further improvement on documentation and code structure.

zhxchen17 commented on torch/_dynamo/functional_export.py:
```diff
@@ -618,6 +619,235 @@ def normalize_graph_module(gm):
+        graph_module._non_persistent_buffers_set = (
+            root._non_persistent_buffers_set.copy()
+        )
+        if sys.version_info >= (3, 14):
```
When we are copying over the module attributes, I want to filter out all the user owned attributes vs torch owned ones. So I use torch.nn.Module annotations (fieldname: type) to filter out these.

zhxchen17 committed:
[export] Use bytecode to flatten graph inputs. (#171110)

Summary:

Previously for dynamo_graph_capture_for_export, we use make_fx() to trace through the dynamo bytecode so that the GraphModule looks like the following:
```
tree_leaves = pytree_flatten(*args)
graph_inputs = self.in_shuffle_graph(*tree_leaves)
graph_outputs = graph(*graph_inputs)
tree_outs = outputs = self.out_shuffle_graph(*graph_outputs)
return pytree.tree_unflatten(tree_outs)
```

After a bunch of experiments, we think the make_fx step is not necessaery, and to further close the gap between strict export and dynamo, we propose the new impl to do the following using torch.fx codegen:

```
graph_inputs = self._dynamo_bytecode_flatten(*args)
graph_outs = graph(*graph_inputs)
return self._dynamo_bytecode_unflatten(*graph_outs)
```
The `_dynamo_bytecode_flatten` and `_dynamo_bytecode_unflatten` wrappers are constructed in the way that the behavior will be 100% identical to directly running bytecode.

This is important for also making precompile reusing the same piece of logic that torch.export uses.

Test Plan: test:test_export -- -r _strict_export_v2

Differential Revision: D89673950
